File: 54. Spiral Matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> res;
        if (matrix.empty())
            return res;
        int li = 0, lj = 0, hi = matrix.size() - 1, hj = matrix[0].size()-1;
        while (li <= hi && lj <= hj) {
            int i = li, j = lj;
            while (j <= hj) {
                res.push_back(matrix[i][j]);
                j++;
            }
            i++;j--;
            while (i <= hi) {
                res.push_back(matrix[i][j]);
                i++;
            }
            i--;j--;
            while (j > lj && li != hi) {
                res.push_back(matrix[i][j]);
                j--;
            }
            
            while (i > li && lj != hj) {
                res.push_back(matrix[i][j]);
                i--;
            }
            
            li++;lj++;hi--;hj--;
        }
        
        return res;
    }
};

File: go/src/leetcode/54. Spiral Matrix/54. Spiral Matrix.go
package main

import "fmt"

func spiralOrder(matrix [][]int) []int {
	if len(matrix) == 0 {
		return []int{}
	}
	m, n := len(matrix), len(matrix[0])
	ans := []int{}
	l1, r1, l2, r2 := 0, m-1, 0, n-1
	for l1<=r1 && l2<=r2 {
		i, j := l1, l2
		for ; j <= r2; j++ {
			ans = append(ans, matrix[i][j])
		}
		i++
		j--
		for ; i <= r1; i++ {
			ans = append(ans, matrix[i][j])
		}
		i--
		j--

		for ; l1 != r1 && j > l2; j-- {
			ans = append(ans, matrix[i][j])
		}

		for ; l2 != r2 && i > l1; i-- {
			ans = append(ans, matrix[i][j])
		}
		l1++
		l2++
		r1--
		r2--
	}
	return ans
}

func main() {
	fmt.Println(spiralOrder([][]int{{1,2,3,4},{5,6,7,8},{9,10,11,12}}))
}
